File: utils/attachments.py
"""Discord 첨부파일 및 커스텀 이모지를 로컬에 다운로드하는 유틸리티."""

import logging
import os
import re
from pathlib import Path

import aiohttp

log = logging.getLogger(__name__)

ATTACHMENTS_DIR = Path(os.getenv("ATTACHMENTS_DIR", "data/attachments"))
EMOJIS_DIR = Path(os.getenv("EMOJIS_DIR", "data/emojis"))

# <:name:id> 또는 <a:name:id> (애니메이션)
CUSTOM_EMOJI_RE = re.compile(r"<(a?):(\w+):(\d+)>")


async def _download_file(url: str, dest: Path) -> bool:
    """URL에서 파일을 다운로드한다. 성공 시 True."""
    dest.parent.mkdir(parents=True, exist_ok=True)
    try:
        async with aiohttp.ClientSession() as session:
            async with session.get(url, timeout=aiohttp.ClientTimeout(total=30)) as resp:
                if resp.status != 200:
                    log.warning("다운로드 실패 (HTTP %s): %s", resp.status, url)
                    return False
                with open(dest, "wb") as f:
                    async for chunk in resp.content.iter_chunked(8192):
                        f.write(chunk)
    except Exception:
        log.exception("다운로드 중 오류: %s", url)
        return False
    return True


async def download_attachment(
    url: str,
    channel_id: str,
    message_id: str,
    filename: str,
) -> str | None:
    """Discord CDN에서 첨부파일을 다운로드하고 로컬 상대 경로를 반환한다."""
    rel_path = f"{channel_id}/{message_id}_{filename}"
    dest = ATTACHMENTS_DIR / rel_path

    if not await _download_file(url, dest):
        return None
    return rel_path


async def download_emojis(content: str) -> None:
    """메시지 텍스트에서 커스텀 이모지를 파싱하고 미저장된 이모지를 다운로드한다."""
    for match in CUSTOM_EMOJI_RE.finditer(content):
        animated, _name, emoji_id = match.group(1), match.group(2), match.group(3)
        ext = "gif" if animated else "png"
        dest = EMOJIS_DIR / f"{emoji_id}.{ext}"

        if dest.exists():
            continue

        url = f"https://cdn.discordapp.com/emojis/{emoji_id}.{ext}"
        await _download_file(url, dest)

File: internal/attachments/attachments.go
// Package attachments downloads Discord CDN content (message attachments
// and custom emoji images) into a local directory. It is the Go translation
// of utils/attachments.py.
package attachments

import (
	"context"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"regexp"
	"time"
)

// CustomEmojiPattern matches Discord's custom-emoji syntax in message text:
//
//	<:name:id>   (static)
//	<a:name:id>  (animated)
var CustomEmojiPattern = regexp.MustCompile(`<(a?):(\w+):(\d+)>`)

// Downloader is a thin wrapper around an HTTP client and the two output
// directories. Construct once at startup and reuse — *http.Client is
// safe for concurrent use.
type Downloader struct {
	AttachmentsDir string
	EmojisDir      string
	HTTP           *http.Client
}

func New(attachmentsDir, emojisDir string) *Downloader {
	return &Downloader{
		AttachmentsDir: attachmentsDir,
		EmojisDir:      emojisDir,
		HTTP:           &http.Client{Timeout: 30 * time.Second},
	}
}

// DownloadAttachment fetches url into <AttachmentsDir>/<channelID>/<messageID>_<filename>
// and returns the relative path (forward slashes, suitable for storing in
// the messages.attachments JSON). On any failure it returns "" — Python
// returns None and the caller stores no local_path; we mirror that.
func (d *Downloader) DownloadAttachment(
	ctx context.Context, url, channelID, messageID, filename string,
) string {
	rel := fmt.Sprintf("%s/%s_%s", channelID, messageID, filename)
	abs := filepath.Join(d.AttachmentsDir, channelID, fmt.Sprintf("%s_%s", messageID, filename))
	if err := d.fetchTo(ctx, url, abs); err != nil {
		return ""
	}
	return rel
}

// DownloadEmojis scans content for custom emoji refs and fetches any not
// already present on disk. Errors are swallowed (best effort) — emoji
// download is purely a UI improvement, not authoritative state.
func (d *Downloader) DownloadEmojis(ctx context.Context, content string) {
	for _, m := range CustomEmojiPattern.FindAllStringSubmatch(content, -1) {
		animated, _, emojiID := m[1], m[2], m[3]
		ext := "png"
		if animated == "a" {
			ext = "gif"
		}
		dest := filepath.Join(d.EmojisDir, fmt.Sprintf("%s.%s", emojiID, ext))
		if _, err := os.Stat(dest); err == nil {
			continue
		}
		url := fmt.Sprintf("https://cdn.discordapp.com/emojis/%s.%s", emojiID, ext)
		_ = d.fetchTo(ctx, url, dest)
	}
}

func (d *Downloader) fetchTo(ctx context.Context, url, dest string) error {
	if err := os.MkdirAll(filepath.Dir(dest), 0o755); err != nil {
		return err
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return err
	}
	resp, err := d.HTTP.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("download %s: status %d", url, resp.StatusCode)
	}
	f, err := os.Create(dest)
	if err != nil {
		return err
	}
	defer f.Close()
	if _, err := io.Copy(f, resp.Body); err != nil {
		return err
	}
	return nil
}
